geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/collections/Bytes2ObjectOpenHashMap.java

```diff
@@ -0,0 +1,1287 @@
+  protected final transient int minN;
+  /** Number of entries in the set */
+  protected int size;
+
+  /**
+   * Creates a new hash map.
+   *
+   * <p>
+   * The actual table size will be the least power of two greater than {@code expected}/{@code f}.
+   *
+   * @param expected the expected number of elements in the hash map.
+   */
+  @SuppressWarnings("unchecked")
+  public Bytes2ObjectOpenHashMap(final int expected) {
+    if (expected < 0) {
+      throw new IllegalArgumentException("The expected number of elements must be non-negative");
+    }
+    minN = n = arraySize(expected, getLoadFactor());
+    mask = n - 1;
+    maxFill = maxFill(n, getLoadFactor());
+    key = new byte[n + 1][];
+    value = (V[]) new Object[n + 1];
+  }
+
+  /**
+   * Creates a new hash map with initial expected {@link Hash#DEFAULT_INITIAL_SIZE} entries
+   * and {@link Hash#DEFAULT_LOAD_FACTOR} as load factor.
+   */
+  public Bytes2ObjectOpenHashMap() {
+    this(DEFAULT_INITIAL_SIZE);
+  }
+
+  /**
+   * Creates a new hash map with {@link Hash#DEFAULT_LOAD_FACTOR} as load factor copying a given
+   * one.
+   *
+   * @param m a {@link Map} to be copied into the new hash map.
+   */
+  public Bytes2ObjectOpenHashMap(final Map<byte[], ? extends V> m) {
+    this(m.size());
+    putAll(m);
+  }
+
+  /**
+   * Returns the hashing strategy.
+   *
+   * @return the hashing strategy of this custom hash map.
+   */
+  public Strategy<byte[]> strategy() {
+    return strategy;
+  }
+
+  private void ensureCapacity(final int capacity) {
+    final int needed = arraySize(capacity, getLoadFactor());
+    if (needed > n) {
+      rehash(needed);
+    }
+  }
+
+  private void tryCapacity(final long capacity) {
+    final int needed = (int) Math.min(1 << 30,
+        Math.max(2, HashCommon.nextPowerOfTwo((long) Math.ceil(capacity / getLoadFactor()))));
+    if (needed > n) {
+      rehash(needed);
+    }
+  }
+
+  private V removeEntry(final int pos) {
+    final V oldValue = value[pos];
+    value[pos] = null;
+    size--;
+    shiftKeys(pos);
+    if (n > minN && size < maxFill / 4 && n > DEFAULT_INITIAL_SIZE) {
+      rehash(n / 2);
+    }
+    return oldValue;
+  }
+
+  @Override
+  public void putAll(Map<? extends byte[], ? extends V> m) {
+    if (getLoadFactor() <= .5) {
+      ensureCapacity(m.size()); // The resulting map will be sized for m.size() elements
+    } else {
+      tryCapacity(size() + m.size()); // The resulting map will be tentatively sized for size() +
+    }
+    // m.size() elements
+    super.putAll(m);
+  }
+
+  private int find(final byte[] k) {
+    byte[] curr;
+    final byte[][] key = this.key;
+    int pos;
+    // The starting point.
+    if ((curr =
+        key[pos = HashCommon.mix(strategy().hashCode(k)) & mask]) == null) {
+      return -(pos + 1);
+    }
+    if (strategy().equals(k, curr)) {
+      return pos;
+    }
+    // There's always an unused entry.
```

Comment:
I don't know but one of the things I removed was support for a "null" key. I wonder if that null key support has something to do with this extra unused entry. But it might be that it just wanted to always have a null element at the beginning of the key and value arrays.